File: Problems/LeetCode/easy/43-move-zeros.js
/*
  
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Example:
    Input: [0,1,0,3,12]
    Output: [1,3,12,0,0]

Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.

*/
let moveZeroes = function(nums) {
    let j = 0;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] !== 0) {
            let tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
            j++;
        }
    }
    return nums;
};

console.log('Example 1: Result - ' + moveZeroes([1,0,1,3,5,7,0,3,12]) + ' Expected:[1,1,3,5,7,3,12,0,0]');

File: Problems/LeetCode/easy/43-move-zeros.py
class Solution(object):
    def moveZeroes(self, nums):
        i = 0
        j = 0
        while i < len(nums):
            if nums[i] != 0:
                tmp = nums[i]
                nums[i] = nums[j]
                nums[j] = tmp
                j += 1
            i += 1
        return nums


instance = Solution()

print(instance.moveZeroes([1,0,1,3,5,7,0,3,12]))